pkg/servicemesh/federation/common/namemapping.go

```diff
@@ -0,0 +1,139 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"maistra.io/api/core/v1alpha1"
+
+	"istio.io/istio/pilot/pkg/model"
+	federationmodel "istio.io/istio/pkg/servicemesh/federation/model"
+)
+
+func setHostname(name *federationmodel.ServiceKey, domainSuffix string) {
+	if name == nil {
+		return
+	}
+	name.Hostname = fmt.Sprintf("%s.%s.%s", name.Name, name.Namespace, domainSuffix)
+}
+
+type NameMapper interface {
+	NameForService(svc *model.Service) *federationmodel.ServiceKey
+	UpdateDefaultMapper(defaults NameMapper)
+}
+
+type nameMatcher struct {
+	domainSuffix string
+	match        v1alpha1.ServiceName
+	alias        *v1alpha1.ServiceName
+}
+
+var _ NameMapper = (*nameMatcher)(nil)
+
+func newNameMatcher(mapping *v1alpha1.ServiceNameMapping, domainSuffix string) NameMapper {
+	var alias *v1alpha1.ServiceName
+	// if it's nil or matches anything, it may as well be nil
+	if mapping.Alias == nil ||
+		((mapping.Alias.Namespace == v1alpha1.MatchAny || mapping.Alias.Namespace == "") &&
+			(mapping.Alias.Name == v1alpha1.MatchAny || mapping.Alias.Name == "")) {
+		alias = nil
+	} else {
+		alias = &v1alpha1.ServiceName{}
+		*alias = *mapping.Alias
```

Comment:
Better use DeepCopy? It would be safer if the api object changes to more complex structures in the future.

Reply:
good call.  i'll update to use `DeepCopy()`